### ISSUE TITLE
Add fused attention and a typed KV cache

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#104]
> Filed against the `performance.yml` template, plus a `feature_request.yml` API section for the KV cache. Subsystem: **CUDA MatMul / Autotuned Kernels**, **Autograd Tape Allocation / Backward Traversal**.

## Performance issue description

`scaled_dot_product_attention` is advertised as `composed` on every backend that has it, and `f32` only:

```
| `scaled_dot_product_attention` | `f32` | `f32` | — | — |
| `bmm`                          | `f32` | `f32` | — | — |
```

On CUDA the composition sits on a hand-written 2D `f32` GEMM whose own module doc says "Unbatched 2D operands only". So one attention call becomes a sequence of separate kernel launches, each writing the full attention matrix to global memory.

**The `[B, H, T, T]` intermediate is the problem.** It is quadratic in sequence length and materialized twice, once forward and once retained for backward. That is what sets the maximum context length a model can train at.

There is also **no KV cache**. Generating `n` tokens re-runs attention over the whole prefix `n` times, making autoregressive inference quadratic where it should be linear.

Neither is a correctness gap, so neither blocks anything. Both are the difference between an implementation that runs and one someone would choose.

## Benchmark plan and measurements

The claim to be demonstrated is memory scaling, not only speed. Measurements to record, before and after, on the #82 runner:

```text
Configuration: B=8, H=32, HEAD_DIM=128, bf16, causal
Hardware: <populate from the #82 runner>

Sequence length | Composed peak mem | Fused peak mem | Composed fwd+bwd | Fused fwd+bwd
----------------|-------------------|----------------|------------------|---------------
 1024           |                   |                |                  |
 2048           |                   |                |                  |
 4096           |                   |                |                  |
 8192           |    OOM expected   |                |   OOM expected   |
16384           |    OOM expected   |                |   OOM expected   |

Expected shape of the result: composed memory grows as O(T^2), fused as O(T).
The crossover length at which fused beats composed on time is a measured number,
not an assumption, and it is what the autotune selection threshold should use.
```

```text
Generation, KV cache:
Prompt 512 tokens, generate 256, batch 1
                     | time/token | total
No cache (today)     |            |        expected: grows linearly with position
With cache           |            |        expected: flat
```

Causal block skipping should be verified by a **benchmark**, not only by output equality: a kernel that computes the upper triangle and then masks it produces identical numbers and roughly twice the work.

## Proposed API sketch

The fused path is a separate catalog operation so a backend advertises it only where it exists, and the composed path stays as the portable fallback:

```rust
// proposed - a distinct catalog operation, not a flag on the existing one
op::FusedAttention
```

The KV cache is a typed structure with a capacity, so overflow is a typed error rather than a reallocation surprise:

```rust
use incin::prelude::*;
type Backend = DefaultBackend;

const MAX_SEQ: usize = 4096;

// proposed
let mut cache: KvCache<s![1, N_KV_HEADS, MAX_SEQ, HEAD_DIM], Backend, f32> = KvCache::new()?;

for step in 0..256 {
    // Attention consumes and extends the cache; only the new token is projected.
    let logits = model.forward_cached(&token, &mut cache)?;
    let next = logits.argmax(1)?;
    token = next;
}

// Capacity is in the type. Appending past it is a typed error, not a realloc.
assert!(matches!(cache.append(&k, &v), Err(Error::CacheCapacityExceeded { .. })));
```

The cache must participate in state traversal correctly: it is **runtime state, not a parameter**, so like RoPE's frequency table in #101 it belongs behind `Buffer` semantics rather than `Param`, and it should be resettable between sequences.

## Shape and type-system invariants

| Invariant | Enforcement |
|---|---|
| Cache capacity is a type-level extent | `KvCache<s![B, H_KV, MAX_SEQ, D], ..>`. Overflow is checked against a static bound. |
| Cache KV head count matches the attention module's | Both derive from the same const parameters as #101, so a mismatch does not compile. |
| Fused and composed produce the same output type | The fused operation is a different `OperationKind` with the same output shape rule, so selection is invisible to the caller's types. |
| Causal masking is a property of the operation | Not a mask tensor passed in, so the kernel can skip blocks. A caller-supplied mask cannot be block-skipped safely in general. |
| Cache is `NoGrad` | Generation is inference. A cache participating in autograd is a bug, and the type should say so. |

## Background

- **Online softmax is the whole technique.** Process the attention matrix in query blocks, keeping a running maximum and normalizer per block, rescaling the accumulated output as the maximum moves. The full `[T, T]` matrix never exists. This is numerically equivalent to the composed path, not an approximation, which is what makes it comparable against the CPU oracle in #83.
- **Backward recomputes rather than stores.** Storing the attention matrix for backward would forfeit the entire memory argument. Recomputation costs an extra forward pass over the attention block and is still dramatically cheaper than the quadratic buffer.
- **Causal skipping halves the work.** With a causal mask, blocks strictly above the diagonal contribute nothing. Skipping them at block granularity is where roughly half the compute disappears for a decoder, and it is only possible when the kernel knows the mask is causal rather than receiving it as data.
- **GQA must be in the same kernel.** #101 makes grouped-query attention a configuration, so a fused kernel handling only full multi-head attention would not serve the module it sits under. With `N_KV_HEADS < N_HEADS`, several query heads read the same KV block, which is a strict improvement for the kernel's memory traffic.
- **Selection must be measured.** Below some sequence length the composed path wins on launch overhead. That crossover is what the existing autotune infrastructure is for, and it should be a recorded measurement rather than a constant someone guessed.

## How this interoperates

- **#85** is a hard dependency twice over: the batched GEMM is what the composed path falls back to, and its cuBLASLt work is where the epilogue and tensor-core machinery gets established.
- **#101** is the module surface this sits under. The fused kernel is selected *inside* `MultiHeadAttention`; a user never chooses it.
- **#83** verifies the fused path against the composed path as oracle, which is exactly the comparison the online-softmax equivalence argument predicts should hold exactly.
- **`crates/incin-backends/src/tuning/`** owns the crossover selection. This is a real call-site for the autotune infrastructure, which `docs/plan/roadmap.md` currently notes as "Infrastructure complete; the call-site API is not exposed."
- **`OPERATION_CATALOG`** gains a `FusedAttention` row; `execution_site` defaults to `Kernel`, so it automatically owes an executor, and the capability table advertises it only on backends that have one.
- **#90** determines whether the fused kernel can run in `bf16` at all, which is the dtype anyone would actually train in.
- **`incin-telemetry`** should report which path was selected per call, otherwise a silent fallback to the composed path looks like the fused kernel simply being slow.

## Acceptance criteria

- Fused attention matches the composed reference within documented tolerances, forward and backward, verified by #83.
- Peak memory for an attention layer is linear in sequence length rather than quadratic, demonstrated by the measurement table above at four sequence lengths.
- The composed path remains available and is selected automatically where the fused kernel is unavailable or unprofitable, with the crossover recorded as a measurement.
- Causal block skipping is verified by a benchmark showing roughly half the work, not only by output equality.
- Grouped-query configurations run in the same kernel.
- Autoregressive generation with a KV cache is flat in time per token, demonstrated by the generation measurement.
- Appending beyond cache capacity is a typed error, covered by a test.
- The selected path is visible in telemetry.

## Alternatives

- **Keep the composed implementation and rely on gradient checkpointing.** Works, and costs a full recomputation of every layer rather than of one attention block. It also leaves the launch overhead and memory traffic in place, which dominate at the short sequence lengths where checkpointing does not help.
- **Materialize the attention matrix in `bf16`** to halve the buffer. Cheap, and still quadratic; it moves the OOM boundary by one power of two and no further.
- **A cache that grows by reallocation** instead of a typed capacity. Friendlier at small scale, and it turns a predictable typed error into an unpredictable allocation stall mid-generation.

## Additional context

Relevant code: `crates/incin-backends/src/cuda/ops/matmul.rs` and `kernels/matmul.cu`, `crates/incin-backends/src/capability/tables.rs`, `crates/incin-core/src/exec/catalog/table.rs`, `crates/incin-backends/src/tuning/`, `crates/incin-core/src/nn/param.rs` (`Buffer`), `docs/benchmarks/`.

Depends on #85, #101, and #90. Verified by #83 on the #82 runner.
